File: src/main/java/fi/aalto/cs/apluscourses/model/Points.java
package fi.aalto.cs.apluscourses.model;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.regex.Pattern;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.json.JSONArray;
import org.json.JSONObject;

public class Points {

  @NotNull
  private static final Pattern submissionIdPattern = Pattern.compile("/submissions/(\\d+)/?$");

  @NotNull
  private final Map<Long, List<Long>> exercises;

  @NotNull
  private final Map<Long, List<Long>> submissions;

  @NotNull
  private final Map<Long, Long> bestSubmissions;

  private Integer exercisesCount;
  private Integer submissionsCount;

  /**
   * Construct an instance with the given maps.
   *
   * @param exercises       A map of exercise group IDs to a list of exercises for that exercise
   *                        group.
   * @param submissions     A map of exercise IDs to a list of submission IDs for that exercise.
   *                        The first element of the list should be the ID of the oldest submission
   *                        and the last element should be the ID of the latest submission.
   * @param bestSubmissions A map of exercise IDs to the IDs of the best submission for each
   *                        exercise.
   */
  public Points(@NotNull Map<Long, List<Long>> exercises,
                @NotNull Map<Long, List<Long>> submissions,
                @NotNull Map<Long, Long> bestSubmissions) {
    this.exercises = exercises;
    this.submissions = submissions;
    this.bestSubmissions = bestSubmissions;
  }

  @NotNull
  public List<Long> getSubmissions(long exerciseId) {
    return submissions.getOrDefault(exerciseId, Collections.emptyList());
  }

  /**
   * Returns the amount of exercises.
   */
  public int getSubmissionsCount() {
    if (submissionsCount == null) {
      submissionsCount = submissions.values().stream().mapToInt(Collection::size).sum();
    }
    return submissionsCount;
  }

  public int getSubmissionsAmount(@NotNull Long id) {
    return submissions.get(id).size();
  }

  @NotNull
  public List<Long> getExercises(long exerciseGroupId) {
    return exercises.getOrDefault(exerciseGroupId, Collections.emptyList());
  }

  /**
   * Returns the amount of exercises.
   */
  public int getExercisesCount() {
    if (exercisesCount == null) {
      exercisesCount = exercises.values().stream().mapToInt(Collection::size).sum();
    }
    return exercisesCount;
  }

  @NotNull
  public Map<Long, Long> getBestSubmissionIds() {
    return Collections.unmodifiableMap(bestSubmissions);
  }

  /**
   * Constructs a {@link fi.aalto.cs.apluscourses.model.exercise.Points} instance from the given JSON object.
   *
   * @param jsonObject The JSON object from which the {@link fi.aalto.cs.apluscourses.model.exercise.Points} instance is constructed.
   */
  @NotNull
  public static fi.aalto.cs.apluscourses.model.exercise.Points fromJsonObject(@NotNull JSONObject jsonObject) {
    JSONArray modulesArray = jsonObject.getJSONArray("modules");
    Map<Long, List<Long>> exercises = new HashMap<>();
    Map<Long, List<Long>> submissions = new HashMap<>();
    Map<Long, Long> bestSubmissions = new HashMap<>();
    for (int i = 0; i < modulesArray.length(); ++i) {
      JSONObject module = modulesArray.getJSONObject(i);
      var exerciseGroupId = module.getLong("id");
      List<Long> exerciseIds = new ArrayList<>();
      JSONArray exercisesArray = module.getJSONArray("exercises");
      for (int j = 0; j < exercisesArray.length(); ++j) {
        JSONObject exercise = exercisesArray.getJSONObject(j);
        long exerciseId = exercise.getLong("id");
        exerciseIds.add(exerciseId);
        parseSubmissions(exercise, exerciseId, submissions);

        var bestSubmissionId = parseSubmissionId(exercise.optString("best_submission"));
        if (bestSubmissionId != null) {
          bestSubmissions.put(exerciseId, bestSubmissionId);
        }
      }
      exercises.put(exerciseGroupId, exerciseIds);
    }
    return new fi.aalto.cs.apluscourses.model.exercise.Points(exercises, submissions, bestSubmissions);
  }

  /*
   * Parses the submissions (IDs and points) from the given JSON and adds them to the given maps.
   */
  private static void parseSubmissions(@NotNull JSONObject exerciseJson,
                                       long exerciseId,
                                       @NotNull Map<Long, List<Long>> submissions) {
    JSONArray submissionsArray = exerciseJson.getJSONArray("submissions_with_points");
    List<Long> submissionIds = new ArrayList<>(submissionsArray.length());
    for (int i = submissionsArray.length() - 1; i >= 0; --i) {
      JSONObject submission = submissionsArray.getJSONObject(i);
      long submissionId = submission.getLong("id");
      submissionIds.add(submissionId);
    }
    submissions.put(exerciseId, submissionIds);
  }

  @Nullable
  private static Long parseSubmissionId(String submissionUrl) {
    var matcher = submissionIdPattern.matcher(submissionUrl);
    if (!matcher.find()) {
      return null;
    }
    return Long.parseLong(matcher.group(1));
  }

}

File: src/main/kotlin/fi/aalto/cs/apluscourses/model/exercise/Points.kt
package fi.aalto.cs.apluscourses.model.exercise

import kotlinx.serialization.Serializable
import java.util.*
import java.util.regex.Pattern

@Serializable
@Suppress("PROVIDED_RUNTIME_TOO_LOW")
class Points
/**
 * Construct an instance with the given maps.
 *
 * @param exercises       A map of exercise group IDs to a list of exercises for that exercise
 * group.
 * @param submissions     A map of exercise IDs to a list of submission IDs for that exercise.
 * The first element of the list should be the ID of the oldest submission
 * and the last element should be the ID of the latest submission.
 * @param bestSubmissions A map of exercise IDs to the IDs of the best submission for each
 * exercise.
 */(
    private val exercises: Map<Long, List<Long>>,
    private val submissions: Map<Long, List<Long>>,
    private val bestSubmissions: Map<Long, Long>
) {
    fun exercisesCount(): Int = exercises.size

    /**
     * Returns the amount of exercises.
     */
//        get() {
//            if (field == null) {
//                field =
//                    exercises.values.stream().mapToInt(ToIntFunction<List<Long>> { obj: List<Long?> -> obj.size }).sum()
//            }
//            return field
//        }
//        private set
    fun submissionsCount(): Int = submissions.size

    /**
     * Returns the amount of exercises.
     */
//        get() {
//            if (field == null) {
//                field = submissions.values.stream().mapToInt(ToIntFunction<List<Long>> { obj: List<Long?> -> obj.size })
//                    .sum()
//            }
//            return field
//        }
//        private set

    fun getSubmissions(exerciseId: Long): List<Long> {
        return submissions.getOrDefault(exerciseId, emptyList())
    }

    fun getSubmissionsAmount(id: Long): Int {
        return submissions[id]!!.size
    }

    fun getExercises(exerciseGroupId: Long): List<Long> {
        return exercises.getOrDefault(exerciseGroupId, emptyList())
    }

    val bestSubmissionIds: Map<Long, Long>
        get() = Collections.unmodifiableMap(bestSubmissions)

    companion object {
        private val submissionIdPattern: Pattern = Pattern.compile("/submissions/(\\d+)/?$")

        /**
         * Constructs a [Points] instance from the given JSON object.
         *
         * @param jsonObject The JSON object from which the [Points] instance is constructed.
         */
//        fun fromJsonObject(jsonObject: JSONObject): Points {
//            val modulesArray: JSONArray = jsonObject.getJSONArray("modules")
//            val exercises: MutableMap<Long, List<Long>> = HashMap()
//            val submissions: MutableMap<Long, List<Long>> = HashMap()
//            val bestSubmissions: MutableMap<Long, Long> = HashMap()
//            for (i in 0 until modulesArray.length()) {
//                val module: JSONObject = modulesArray.getJSONObject(i)
//                val exerciseGroupId: Unit = module.getLong("id")
//                val exerciseIds: MutableList<Long> = ArrayList()
//                val exercisesArray: JSONArray = module.getJSONArray("exercises")
//                for (j in 0 until exercisesArray.length()) {
//                    val exercise: JSONObject = exercisesArray.getJSONObject(j)
//                    val exerciseId: Long = exercise.getLong("id")
//                    exerciseIds.add(exerciseId)
//                    parseSubmissions(exercise, exerciseId, submissions)
//
//                    val bestSubmissionId = parseSubmissionId(exercise.optString("best_submission"))
//                    if (bestSubmissionId != null) {
//                        bestSubmissions[exerciseId] = bestSubmissionId
//                    }
//                }
//                exercises[exerciseGroupId] = exerciseIds
//            }
//            return Points(exercises, submissions, bestSubmissions)
//        }

        /*
   * Parses the submissions (IDs and points) from the given JSON and adds them to the given maps.
   */
//        private fun parseSubmissions(
//            exerciseJson: JSONObject,
//            exerciseId: Long,
//            submissions: MutableMap<Long, List<Long>>
//        ) {
//            val submissionsArray: JSONArray = exerciseJson.getJSONArray("submissions_with_points")
//            val submissionIds: MutableList<Long> = ArrayList<Any?>(submissionsArray.length())
//            for (i in submissionsArray.length() - 1 downTo 0) {
//                val submission: JSONObject = submissionsArray.getJSONObject(i)
//                val submissionId: Long = submission.getLong("id")
//                submissionIds.add(submissionId)
//            }
//            submissions[exerciseId] = submissionIds
//        }

        private fun parseSubmissionId(submissionUrl: String): Long? {
            val matcher = submissionIdPattern.matcher(submissionUrl)
            if (!matcher.find()) {
                return null
            }
            return matcher.group(1).toLong()
        }
    }
}
